### PR TITLE
Provide a new executable to rule them all

### PR DESCRIPTION
**Pull request by hasegeli**

### CODE REVIEW

File: adminapi/cli.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            raise Exception('Multiple servers')
+
+    changes = bool(args.reset or args.update)
+    for server in servers:
+        print(server)
+        if changes:
+            apply_changes(server, args.reset, args.update)
+        print_server(server, args.attrs)
+    if changes:
+        servers.commit()
```
Comment (TheFRedFox) [inline adminapi/cli.py]: Will it print errors?

File: adminapi/cli.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                '.innogames.net',
+            ]):
+                value += '.ig.local'
+
+        if value in [None, True, False]:
+            value = '{{{}}}'.format(str(value).lower())
+
+        values.append(value)
+
+    print(*values, sep='\t')
```
Comment (TheFRedFox) [inline adminapi/cli.py]: It seems you just print the values. Wouldn't it be helpful to also print the related attribute names?

File: adminapi/cli.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        )
+        .restrict(*(args.attrs + args.reset + [k for k, v in args.update]))
+        .order_by(*args.orderby)
+    )
+
+    if args.one:
+        if not servers:
+            raise Exception('No servers')
+        if len(servers) > 1:
+            raise Exception('Multiple servers')
```
[the remark this answers is not in the document]
Comment (TheFRedFox) [reply adminapi/cli.py]: Maybe you can/wanna combine it to something like:
``` python
if len(servers) != 1:
    raise Exception('Expected exactly one server, but found {} servers'.format(len(servers)))
```

File: adminapi/cli.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+from adminapi import _api_settings
+from adminapi.dataset import QuerySet
+from adminapi.dataset.base import MultiAttr
+from adminapi.utils.parse import parse_query
+
+
+def parse_args():
+    parser = ArgumentParser('igserver')
+    parser.add_argument('quals', nargs='+')
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply adminapi/cli.py]: Renaming to `query`.

File: adminapi/cli.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    parser.add_argument(
+        '-a',
+        '--attrs',
+        nargs='+',
+        default=['hostname'],
+        help='The attributes to fetch',
+    )
+    parser.add_argument(
+        '-o',
+        '--orderby',
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply adminapi/cli.py]: Renaming to `order` only...

File: adminapi/cli.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        nargs='+',
+        default=['hostname'],
+        help='Attributes to order by the result',
+    )
+    parser.add_argument(
+        '-r',
+        '--reset',
+        nargs='+',
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply adminapi/cli.py]: Yes, it means "if the argument is used, then at least one value must be provided".  `required` controls, if the arguments must be used.  By default, the ones start with `-` or `--` are not required.

File: adminapi/cli.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hasegeli) [reply adminapi/cli.py]: Yes, I intended this as a security feature.  I am rewording the message as suggested.

File: adminapi/cli.py (hunk 1 above)
Comment (hasegeli) [reply adminapi/cli.py]: It wouldn't `print` anything, but raise some kind of `Exception`.

Comment (TheFRedFox) [reply adminapi/cli.py]: Yeah, that's fair enough I guess. Or do you also want to notify the user if the changes actually did work at the end?

File: adminapi/cli.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheFRedFox) [reply adminapi/cli.py]: Maybe you wanna clarify the function of the argument then in the `help` message, too.

File: adminapi/cli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheFRedFox) [reply adminapi/cli.py]: Got it, then nevermind. 🙂

File: adminapi/cli.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        server[attribute_id].clear()
+    for attribute_id, value in update:
+        if type(server[attribute_id]) is MultiAttr:
+            server[attribute_id].add(value)
+        elif type(server[attribute_id]) is bool:
+            server[attribute_id] = strtobool(value)
+        elif type(server[attribute_id]) is int:
+            server[attribute_id] = int(value)
+        else:
+            server[attribute_id] = value
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply adminapi/cli.py]: Good idea.  I am making it a method of `Server`.

File: adminapi/cli.py (hunk 2 above)
Comment (hasegeli) [reply adminapi/cli.py]: Not by default.  The user already knows what attributes they required.  Also, I am trying to keep this compatible with the existing executable.  Typical usage is something like `gsh $(igserver project=foe function=db)`.

File: run.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env python3
+"""igserver - Convenience script for testing
+
+Copyright (c) 2017, InnoGames GmbH
+"""
+
+from adminapi.cli import main
+
+main()
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply run.py]: I don't think we need on this one.  This script has nothing else.

File: adminapi/cli.py (hunk 1 above)
Comment (hasegeli) [reply adminapi/cli.py]: I take the Unix approach of not emitting anything when everything is fine.

Comment (TheFRedFox) [reply adminapi/cli.py]: Alright then. 👍

File: adminapi/cli.py (hunk 2 above)
Comment (TheFRedFox) [reply adminapi/cli.py]: Fair enough then.

File: run.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (TheFRedFox) [reply run.py]: Fair enough.

**Pull request merged**